File: baekjoon/divide_and_conquer/colorpaper.py
import sys
input = sys.stdin.readline

def paper(board,row,col,length,answer):
	if length == 1:
		return board[row][col]
	else:
		lu = paper(board,row,col,int(length/2),answer)
		ld = paper(board,row+int(length/2),col,int(length/2),answer)
		ru = paper(board,row,col+int(length/2),int(length/2),answer)
		rd = paper(board,row+int(length/2),col+int(length/2),int(length/2),answer)
		if lu == ld == ru == rd == 1:
			return 1
		elif lu == ld == ru == rd == 0:
			return 0
		else:
			if lu == 1:
				answer[1] += 1
			elif lu == 0:
				answer[0] += 1
			if ld == 1:
				answer[1] += 1
			elif ld == 0:
				answer[0] += 1
			if ru == 1:
				answer[1] += 1
			elif ru == 0:
				answer[0] += 1
			if rd == 1:
				answer[1] += 1
			elif rd == 0:
				answer[0] += 1
	return answer


def solution(board):
	answer = [0,0]
	answer = paper(board,0,0,len(board),answer)
	if answer == 1:
		print(0)
		print(1)
	elif answer == 0:
		print(1)
		print(0)
	else:
		print(answer[0])
		print(answer[1])

if __name__ == "__main__":
	N = int(input())
	board = []
	for _ in range(N):
		line = list(map(int,input().split()))
		board.append(line)
	solution(board)

File: baekjoon/divide_and_conquer/colorpaper.java
import java.util.*;
import java.io.*;

public class Main{

	static int N;
	static int[][] paper;
	static int white,blue = 0;

	static int divide(int length,int row,int col){
		if(length == 1){
			return paper[row][col];
		}
		else{
			int ul = divide(length/2,row,col);
			int ur = divide(length/2,row,col+length/2);
			int dl = divide(length/2,row+length/2,col);
			int dr = divide(length/2,row+length/2,col+length/2);
			if(ul == 0 && ur == 0 && dl == 0 && dr == 0){
				return 0;
			}
			else if(ul == 1 && ur == 1 && dl == 1 && dr == 1){
				return 1;
			}
			else{
				if(ul == 0){
					white++;
				}
				else if(ul == 1){
					blue++;
				}
				if(ur == 0){
					white++;
				}
				else if(ur == 1){
					blue++;
				}
				if(dl == 0){
					white++;
				}
				else if(dl == 1){
					blue++;
				}
				if(dr == 0){
					white++;
				}
				else if(dr == 1){
					blue++;
				}
			}
		}
		return -1;
	}

	static void solution(){
		int result = divide(N,0,0);
		if(result == 0){
			white++;
		}else if(result == 1){
			blue++;
		}
		System.out.println(white);
		System.out.println(blue);
	}


	public static void main(String args[]) throws Exception{
		BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
		StringTokenizer st = new StringTokenizer(br.readLine());
		N = Integer.parseInt(st.nextToken());
		paper = new int[N][N];
		for(int i=0;i<N;i++){
			st = new StringTokenizer(br.readLine());
			for(int j=0;j<N;j++){
				paper[i][j] = Integer.parseInt(st.nextToken());
			}
		}
		solution();
	}
}